File: graph/Dijkstra.java
/**
 * Dijkstra.
 *
 * http://www.geeksforgeeks.org/greedy-algorithms-set-6-dijkstras-shortest-path-algorithm/
 */

import java.util.Arrays;

public class Dijkstra {

  public static void main(String[] args) {
    test();
  }

  private static void test() {
    int[][] graph = {
      {0, 4, 0, 0, 0, 0, 0, 8, 0},
      {4, 0, 8, 0, 0, 0, 0, 11, 0},
      {0, 8, 0, 7, 0, 4, 0, 0, 2},
      {0, 0, 7, 0, 9, 14, 0, 0, 0},
      {0, 0, 0, 9, 0, 10, 0, 0, 0},
      {0, 0, 4, 0, 10, 0, 2, 0, 0},
      {0, 0, 0, 14, 0, 2, 0, 1, 6},
      {8, 11, 0, 0, 0, 0, 1, 0, 7},
      {0, 0, 2, 0, 0, 0, 6, 7, 0}
    };
    dijkstra(graph, 0);
  }

  public static void dijkstra(int[][] graph, int src) {
    int len = graph.length;
    // visited[i] represents is i has been treated as local min.
    boolean[] visited = new boolean[len];

    // Init dist.
    int[] dist = new int[len];
    Arrays.fill(dist, Integer.MAX_VALUE);
    dist[src] = 0;

    for (int count = 0; count < len - 1; ++count) {
      // Get the min dist of cur status.
      int u = minDistance(dist, visited);
      visited[u] = true;

      /*
       * Not sure this part of code will be hit or not.
       * Reason: in method minDistance the if condition "... && dist[v] < min"
       * will never be true if dist[v] == MAX_VALUE because init value of min
       * itself is MAX_VALUE.
       *
       * if (dist[u] == Integer.MAX_VALUE) {
       *   continue;
       * }
       */

      // If we've known the cost from src to u, update if from src to u then
      // to v costs less than known dist[u].
      for (int v = 0; v < len; ++v) {
        if (!visited[v] && graph[u][v] != 0) {
          // update local min.
          dist[v] = Math.min(dist[v], dist[u] + graph[u][v]);
        }
      }
    }

    System.out.println("Vertes Distance from Source " + src);
    for (int i = 0; i < len; ++i) {
      System.out.println(i + "\t\t" + dist[i]);
    }
  }

  /**
   * Given the cur dijkstra status, find the min distance of cur status.
   */
  public static int minDistance(int[] dist, boolean[] visited) {
    int min = Integer.MAX_VALUE;
    int min_index = 0;

    for (int v = 0; v < dist.length; ++v) {
      if (!visited[v] && dist[v] < min) {
        min = dist[v];
        min_index = v;
      }
    }

    return min_index;
  }
}

File: graph/dijkstra.py
""" dijkstra.py


"""

import heapq
import sys


def dijkstra_priority_queue(graph, src):
    """Optimized version of dijkstra using priority queue.

    Time complexity: O( (E + V)logV), E stands for edge and V for vertices.
    """
    length = len(graph)

    visited = set()
    dist = [sys.maxint for _ in xrange(length)]
    previous = [sys.maxint for _ in xrange(length)]
    # The dist from src to src is 0 and the first time to call min_dist will
    # return src.
    dist[src] = 0

    heap = [(0, src)] # Weight, index

    while heap:
        (_, min_idx) = heapq.heappop(heap)
        if min_idx in visited:
            continue

        visited.add(min_idx)

        for i in xrange(length):
            if i not in visited and \
                graph[min_idx][i] != 0 and \
                dist[min_idx] != sys.maxint and \
                dist[min_idx] + graph[min_idx][i] < dist[i]:
                    dist[i] = dist[min_idx] + graph[min_idx][i]
                    heapq.heappush(heap, (dist[i], i))

    print 'Min vertes distance from src', src
    for i, val in enumerate(dist):
        print i, '\t', val


def dijkstra(graph, src):
    """Dijkstra algorithm for shortest path.

    Maintaining an array dist[i], representing known shortest path from src
    to i, with initial dist[src] = 0 (the shortest path to itself is zero).
    For each iteration, get the shortest path so far from src and try to update
    known path if taking shortest path is better:
        dist[min] + graph[min][i] < dist[i]

    Edge weights must be non-negative, so it can be assumed that taking less edges
    are always better.
    Time complexity O(V^2), V stands for vertices.

    :param graph: 2d list of int. Adjacency matrix of graph.
    :param src: int. source index.
    """
    length = len(graph)

    visited = set()
    dist = [sys.maxint for _ in xrange(length)]
    # The dist from src to src is 0 and the first time to call min_dist will
    # return src.
    dist[src] = 0

    for _ in xrange(length):
        min_idx = min_dist(dist, visited)
        visited.add(min_idx)
        for i in xrange(length):
            # Do not travel back to visited vertex.
            if i not in visited and \
                graph[min_idx][i] != 0 and \
                dist[min_idx] != sys.maxint and \
                dist[min_idx] + graph[min_idx][i] < dist[i]:
                # min_idx and i are connected.
                # shortest path from src to min_idx is found.
                # from src to min_idx and min_idx to i is shorter.
                dist[i] = dist[min_idx] + graph[min_idx][i]

    print 'Min vertes distance from src', src
    for i, val in enumerate(dist):
        print i, '\t', val


def min_dist(dist, visited):
    """"""
    local_min = sys.maxint
    min_idx = 0
    for i, val in enumerate(dist):
        if i not in visited and val < local_min:
            local_min = val
            min_idx = i

    return min_idx


def main():
    #graph = [[1, 2, 3], [1, 2, 3], [1, 2, 3]]
    graph = [[0, 4, 0, 0, 0, 0, 0, 8, 0],
             [4, 0, 8, 0, 0, 0, 0, 11, 0],
             [0, 8, 0, 7, 0, 4, 0, 0, 2],
             [0, 0, 7, 0, 9, 14, 0, 0, 0],
             [0, 0, 0, 9, 0, 10, 0, 0, 0],
             [0, 0, 4, 0, 10, 0, 2, 0, 0],
             [0, 0, 0, 14, 0, 2, 0, 1, 6],
             [8, 11, 0, 0, 0, 0, 1, 0, 7],
             [0, 0, 2, 0, 0, 0, 6, 7, 0]]


    #for i in xrange(len(graph)):
    dijkstra_priority_queue(graph, 0)
    dijkstra(graph, 0)


if __name__ == '__main__':
    main()
